File: bid754-codec-py/benchmarks/bench_runner.py
"""Standalone BID codec benchmark leg (Python, stdlib-only).

Operands come from the hand-pinned shared contract
``bid754-codec-go/testdata/codec_benchmark_operands.json``, consumed by all
four codec benchmark legs (Go ``testing.B``, Rust criterion, the JS script,
and this script). Setup re-checks the contract's exactness round trips
(``encode(decode(bits)) == bits``, ``to_string(decode(bits)) ==
decimal_string``, ``encode(from_string(string)) == bits``) and raises on any
violation, so the run rejects invalid or inexact operands instead of timing
them. Every row rotates operands with an ``i % n`` counter and feeds results
into an observable sink printed at the end, matching the other legs.
Interpreted-leg caveat: each timed iteration pays a fixed harness dispatch
cost (one callable call plus one sink call), uniform across rows, so
within-leg comparisons are sound but absolute ns/op includes that constant;
prefer the compiled legs for cross-language absolute comparisons.
Benchmark infrastructure, not a regular verification domain.

Run from the ``bid754-codec-py`` package directory:
``python3 benchmarks/bench_runner.py [samples]``.
"""

from __future__ import annotations

import json
import re
import sys
import time
from dataclasses import dataclass
from pathlib import Path

_PACKAGE_ROOT = Path(__file__).resolve().parent.parent
sys.path.insert(0, str(_PACKAGE_ROOT))

from bid_codec import (  # noqa: E402
    Components,
    Kind,
    decode32,
    decode64,
    decode128,
    encode32,
    encode64,
    encode128,
    from_string,
    to_string,
)

_OPERAND_PATH = (
    _PACKAGE_ROOT.parent / "bid754-codec-go" / "testdata" / "codec_benchmark_operands.json"
)

_HEX_RE = re.compile(r"^[0-9a-f]+$")


def _parse_hex_word(width: str, index: int, field: str, hex_text: object, bits: int) -> int:
    digits = bits // 4
    if (
        not isinstance(hex_text, str)
        or len(hex_text) != digits
        or not _HEX_RE.fullmatch(hex_text)
    ):
        raise ValueError(
            f"{width}[{index}].{field}: hex {hex_text!r} must be exactly {digits} lowercase hex digits"
        )
    return int(hex_text, 16)


@dataclass
class _WordOperand:
    """Single-word operand (decimal32 and decimal64)."""

    bits: int
    comp: Components
    string: str


@dataclass
class _Operand128:
    lo: int
    hi: int
    comp: Components
    string: str


def _load_operands() -> tuple[list[_WordOperand], list[_WordOperand], list[_Operand128]]:
    file = json.loads(_OPERAND_PATH.read_text(encoding="utf-8"))
    if file["format_version"] != 1:
        raise ValueError(f"benchmark operand format_version {file['format_version']}, want 1")
    for width in ("decimal32", "decimal64", "decimal128"):
        if not file.get(width):
            raise ValueError(f"benchmark operand contract requires non-empty {width} entries")

    operands32: list[_WordOperand] = []
    for i, entry in enumerate(file["decimal32"]):
        if entry.get("hex_hi") is not None:
            raise ValueError(f"decimal32[{i}]: hex_hi is only valid for decimal128 entries")
        bits = _parse_hex_word("decimal32", i, "hex", entry["hex"], 32)
        comp = decode32(bits)
        if comp.kind is not Kind.NORMAL:
            raise ValueError(f"decimal32[{i}] must decode NORMAL")
        if encode32(comp) != bits:
            raise ValueError(f"decimal32[{i}] operand is not canonical")
        if to_string(comp) != entry["decimal_string"]:
            raise ValueError(f"decimal32[{i}] canonical string mismatch")
        if encode32(from_string(entry["decimal_string"])) != bits:
            raise ValueError(f"decimal32[{i}] string operand is not exactly representable")
        operands32.append(_WordOperand(bits=bits, comp=comp, string=entry["decimal_string"]))

    operands64: list[_WordOperand] = []
    for i, entry in enumerate(file["decimal64"]):
        if entry.get("hex_hi") is not None:
            raise ValueError(f"decimal64[{i}]: hex_hi is only valid for decimal128 entries")
        bits = _parse_hex_word("decimal64", i, "hex", entry["hex"], 64)
        comp = decode64(bits)
        if comp.kind is not Kind.NORMAL:
            raise ValueError(f"decimal64[{i}] must decode NORMAL")
        if encode64(comp) != bits:
            raise ValueError(f"decimal64[{i}] operand is not canonical")
        if to_string(comp) != entry["decimal_string"]:
            raise ValueError(f"decimal64[{i}] canonical string mismatch")
        if encode64(from_string(entry["decimal_string"])) != bits:
            raise ValueError(f"decimal64[{i}] string operand is not exactly representable")
        operands64.append(_WordOperand(bits=bits, comp=comp, string=entry["decimal_string"]))

    operands128: list[_Operand128] = []
    for i, entry in enumerate(file["decimal128"]):
        lo = _parse_hex_word("decimal128", i, "hex", entry["hex"], 64)
        hi = _parse_hex_word("decimal128", i, "hex_hi", entry.get("hex_hi"), 64)
        comp = decode128(lo, hi)
        if comp.kind is not Kind.NORMAL:
            raise ValueError(f"decimal128[{i}] must decode NORMAL")
        if encode128(comp) != (lo, hi):
            raise ValueError(f"decimal128[{i}] operand is not canonical")
        if to_string(comp) != entry["decimal_string"]:
            raise ValueError(f"decimal128[{i}] canonical string mismatch")
        if encode128(from_string(entry["decimal_string"])) != (lo, hi):
            raise ValueError(f"decimal128[{i}] string operand is not exactly representable")
        operands128.append(_Operand128(lo=lo, hi=hi, comp=comp, string=entry["decimal_string"]))

    return operands32, operands64, operands128


_SINK = 0


def _sink_components(comp: Components) -> None:
    global _SINK
    _SINK ^= comp.coefficient ^ comp.payload ^ comp.exponent ^ comp.kind.value ^ int(comp.sign)


def _sink_string(s: str) -> None:
    global _SINK
    _SINK ^= len(s)


def _sink_int(v: int) -> None:
    global _SINK
    _SINK ^= v


_TARGET_SAMPLE_NS = 100_000_000  # ~100ms per timed sample
_CALIBRATION_ITERS = 2_000


def _run_batch(fn, iters: int) -> int:
    start = time.perf_counter_ns()
    for i in range(iters):
        fn(i)
    return time.perf_counter_ns() - start


def _bench_row(name: str, samples: int, fn) -> None:
    calibrated = max(_run_batch(fn, _CALIBRATION_ITERS), 1)
    iters = (_TARGET_SAMPLE_NS * _CALIBRATION_ITERS) // calibrated
    iters = max(1_000, min(5_000_000, iters))
    measured: list[float] = []
    for _ in range(samples):
        elapsed = _run_batch(fn, iters)
        measured.append(elapsed / iters)
    median = sorted(measured)[len(measured) // 2]
    rendered = ",".join(f"{v:.1f}" for v in measured)
    print(f"BENCH {name} ns_op_median={median:.1f} iters={iters} samples_ns_op=[{rendered}]")


def main() -> None:
    samples = int(sys.argv[1]) if len(sys.argv) > 1 else 5
    operands32, operands64, operands128 = _load_operands()
    n32, n64, n128 = len(operands32), len(operands64), len(operands128)
    print(
        f"BENCH-LEG codec-py python={sys.version_info.major}.{sys.version_info.minor}."
        f"{sys.version_info.micro} samples={samples} operands={n32}/{n64}/{n128}"
    )

    _bench_row(
        "codec_bid32/decode", samples, lambda i: _sink_components(decode32(operands32[i % n32].bits))
    )
    _bench_row(
        "codec_bid32/encode", samples, lambda i: _sink_int(encode32(operands32[i % n32].comp))
    )
    _bench_row(
        "codec_bid32/to_string", samples, lambda i: _sink_string(to_string(operands32[i % n32].comp))
    )
    _bench_row(
        "codec_bid32/from_string",
        samples,
        lambda i: _sink_components(from_string(operands32[i % n32].string)),
    )

    _bench_row(
        "codec_bid64/decode", samples, lambda i: _sink_components(decode64(operands64[i % n64].bits))
    )
    _bench_row(
        "codec_bid64/encode", samples, lambda i: _sink_int(encode64(operands64[i % n64].comp))
    )
    _bench_row(
        "codec_bid64/to_string", samples, lambda i: _sink_string(to_string(operands64[i % n64].comp))
    )
    _bench_row(
        "codec_bid64/from_string",
        samples,
        lambda i: _sink_components(from_string(operands64[i % n64].string)),
    )

    _bench_row(
        "codec_bid128/decode",
        samples,
        lambda i: _sink_components(decode128(operands128[i % n128].lo, operands128[i % n128].hi)),
    )

    def _encode128_row(i: int) -> None:
        lo, hi = encode128(operands128[i % n128].comp)
        _sink_int(lo ^ hi)

    _bench_row("codec_bid128/encode", samples, _encode128_row)
    _bench_row(
        "codec_bid128/to_string",
        samples,
        lambda i: _sink_string(to_string(operands128[i % n128].comp)),
    )
    _bench_row(
        "codec_bid128/from_string",
        samples,
        lambda i: _sink_components(from_string(operands128[i % n128].string)),
    )

    print(f"BENCH-SINK {_SINK}")


if __name__ == "__main__":
    main()

File: bid754-codec-js/bench_runner.mjs
// Standalone BID codec benchmark leg (JavaScript, dependency-free).
//
// Operands come from the hand-pinned shared contract
// ../bid754-codec-go/testdata/codec_benchmark_operands.json, consumed by all
// four codec benchmark legs (Go testing.B, Rust criterion, this script, and
// bid754-codec-py/benchmarks/bench_runner.py). Setup re-checks the contract's
// exactness round trips (encode(decode(bits)) == bits, toString(decode(bits))
// == decimal_string, encode(fromString(string)) == bits) and throws on any
// violation, so the run rejects invalid or inexact operands instead of timing
// them. Every row rotates operands with an i % n counter and feeds results
// into an observable sink printed at the end, matching the other legs.
// Interpreted-leg caveat: each timed iteration pays a fixed harness dispatch
// cost (one closure call plus the sink update), uniform across rows, so
// within-leg comparisons are sound but absolute ns/op includes that constant;
// prefer the compiled legs for cross-language absolute comparisons.
// Requires the built package (npm run build); imports resolve through the
// package self-reference. Benchmark infrastructure, not a regular
// verification domain.
import { readFileSync } from "node:fs";
import {
  Kind,
  decode32,
  decode64,
  decode128,
  encode32,
  encode64,
  encode128,
  fromString,
  toString,
} from "bid754-codec";

const operandPath = new URL(
  "../bid754-codec-go/testdata/codec_benchmark_operands.json",
  import.meta.url,
);
const file = JSON.parse(readFileSync(operandPath, "utf8"));
if (file.format_version !== 1) {
  throw new Error(`benchmark operand format_version ${file.format_version}, want 1`);
}
for (const width of ["decimal32", "decimal64", "decimal128"]) {
  if (!Array.isArray(file[width]) || file[width].length === 0) {
    throw new Error(`benchmark operand contract requires non-empty ${width} entries`);
  }
}

function parseHexWord(width, index, field, hex, bits) {
  const digits = bits / 4;
  if (typeof hex !== "string" || hex.length !== digits || !/^[0-9a-f]+$/.test(hex)) {
    throw new Error(`${width}[${index}].${field}: hex ${JSON.stringify(hex)} must be exactly ${digits} lowercase hex digits`);
  }
  return BigInt(`0x${hex}`);
}

const operands32 = file.decimal32.map((entry, i) => {
  if (entry.hex_hi !== undefined && entry.hex_hi !== null) {
    throw new Error(`decimal32[${i}]: hex_hi is only valid for decimal128 entries`);
  }
  const bits = Number(parseHexWord("decimal32", i, "hex", entry.hex, 32));
  const comp = decode32(bits);
  if (comp.kind !== Kind.Normal) {
    throw new Error(`decimal32[${i}] must decode Normal`);
  }
  if (encode32(comp) !== bits) {
    throw new Error(`decimal32[${i}] operand is not canonical`);
  }
  if (toString(comp) !== entry.decimal_string) {
    throw new Error(`decimal32[${i}] canonical string mismatch`);
  }
  if (encode32(fromString(entry.decimal_string)) !== bits) {
    throw new Error(`decimal32[${i}] string operand is not exactly representable`);
  }
  return { bits, comp, string: entry.decimal_string };
});

const operands64 = file.decimal64.map((entry, i) => {
  if (entry.hex_hi !== undefined && entry.hex_hi !== null) {
    throw new Error(`decimal64[${i}]: hex_hi is only valid for decimal128 entries`);
  }
  const bits = parseHexWord("decimal64", i, "hex", entry.hex, 64);
  const comp = decode64(bits);
  if (comp.kind !== Kind.Normal) {
    throw new Error(`decimal64[${i}] must decode Normal`);
  }
  if (encode64(comp) !== bits) {
    throw new Error(`decimal64[${i}] operand is not canonical`);
  }
  if (toString(comp) !== entry.decimal_string) {
    throw new Error(`decimal64[${i}] canonical string mismatch`);
  }
  if (encode64(fromString(entry.decimal_string)) !== bits) {
    throw new Error(`decimal64[${i}] string operand is not exactly representable`);
  }
  return { bits, comp, string: entry.decimal_string };
});

const operands128 = file.decimal128.map((entry, i) => {
  const lo = parseHexWord("decimal128", i, "hex", entry.hex, 64);
  const hi = parseHexWord("decimal128", i, "hex_hi", entry.hex_hi, 64);
  const comp = decode128(lo, hi);
  if (comp.kind !== Kind.Normal) {
    throw new Error(`decimal128[${i}] must decode Normal`);
  }
  const [encodedLo, encodedHi] = encode128(comp);
  if (encodedLo !== lo || encodedHi !== hi) {
    throw new Error(`decimal128[${i}] operand is not canonical`);
  }
  if (toString(comp) !== entry.decimal_string) {
    throw new Error(`decimal128[${i}] canonical string mismatch`);
  }
  const [reLo, reHi] = encode128(fromString(entry.decimal_string));
  if (reLo !== lo || reHi !== hi) {
    throw new Error(`decimal128[${i}] string operand is not exactly representable`);
  }
  return { lo, hi, comp, string: entry.decimal_string };
});

// Two sinks so the timed loops never pay Number<->BigInt conversion cost:
// bigint results XOR into sinkBig, small numeric observations XOR into
// sinkNum. Both are printed at the end so every result stays observable.
let sinkBig = 0n;
let sinkNum = 0;

function sinkComponents(comp) {
  sinkBig ^= comp.coefficient ^ comp.payload;
  sinkNum ^= comp.exponent ^ comp.kind ^ (comp.sign ? 1 : 0);
}

function sinkString(s) {
  sinkNum ^= s.length;
}

const SAMPLES = Number.parseInt(process.argv[2] ?? "5", 10);
const TARGET_SAMPLE_NS = 100_000_000n; // ~100ms per timed sample
const CALIBRATION_ITERS = 2_000;

function runBatch(fn, iters) {
  const start = process.hrtime.bigint();
  for (let i = 0; i < iters; i++) {
    fn(i);
  }
  return process.hrtime.bigint() - start;
}

function benchRow(name, fn) {
  // Warmup plus calibration: size the timed batch to ~TARGET_SAMPLE_NS.
  let calibrated = runBatch(fn, CALIBRATION_ITERS);
  if (calibrated <= 0n) {
    calibrated = 1n;
  }
  let iters = Number((TARGET_SAMPLE_NS * BigInt(CALIBRATION_ITERS)) / calibrated);
  iters = Math.max(1_000, Math.min(20_000_000, iters));
  const samples = [];
  for (let s = 0; s < SAMPLES; s++) {
    const elapsed = runBatch(fn, iters);
    samples.push(Number(elapsed) / iters);
  }
  const sorted = [...samples].sort((a, b) => a - b);
  const median = sorted[Math.floor(sorted.length / 2)];
  const rendered = samples.map((v) => v.toFixed(1)).join(",");
  console.log(`BENCH ${name} ns_op_median=${median.toFixed(1)} iters=${iters} samples_ns_op=[${rendered}]`);
}

const n32 = operands32.length;
const n64 = operands64.length;
const n128 = operands128.length;

console.log(`BENCH-LEG codec-js node=${process.version} samples=${SAMPLES} operands=${n32}/${n64}/${n128}`);

benchRow("codec_bid32/decode", (i) => {
  sinkComponents(decode32(operands32[i % n32].bits));
});
benchRow("codec_bid32/encode", (i) => {
  sinkNum ^= encode32(operands32[i % n32].comp);
});
benchRow("codec_bid32/to_string", (i) => {
  sinkString(toString(operands32[i % n32].comp));
});
benchRow("codec_bid32/from_string", (i) => {
  sinkComponents(fromString(operands32[i % n32].string));
});

benchRow("codec_bid64/decode", (i) => {
  sinkComponents(decode64(operands64[i % n64].bits));
});
benchRow("codec_bid64/encode", (i) => {
  sinkBig ^= encode64(operands64[i % n64].comp);
});
benchRow("codec_bid64/to_string", (i) => {
  sinkString(toString(operands64[i % n64].comp));
});
benchRow("codec_bid64/from_string", (i) => {
  sinkComponents(fromString(operands64[i % n64].string));
});

benchRow("codec_bid128/decode", (i) => {
  sinkComponents(decode128(operands128[i % n128].lo, operands128[i % n128].hi));
});
benchRow("codec_bid128/encode", (i) => {
  const [lo, hi] = encode128(operands128[i % n128].comp);
  sinkBig ^= lo ^ hi;
});
benchRow("codec_bid128/to_string", (i) => {
  sinkString(toString(operands128[i % n128].comp));
});
benchRow("codec_bid128/from_string", (i) => {
  sinkComponents(fromString(operands128[i % n128].string));
});

console.log(`BENCH-SINK ${sinkBig}/${sinkNum}`);
